### PR TITLE
UP-4727 : Managing uPortal running on several servername with CAS

### PR DESCRIPTION
**Pull request by jgribonvald**
##### Checklist

-   [*] the [individual contributor license agreement][] is signed
-   [*] commit message follows [commit guidelines][]
-   [*] tests are included
-   [ ] documentation is changed or added
-   [ ] [message properties][] have been updated with new phrases
-   [ ] view conforms with [WCAG 2.0 AA][]

##### Description of change

Upgrade from the previous Pull Request: https://github.com/Jasig/uPortal/pull/689

This PR is needed when you are deploying a multi-tenant uPortal instance working on several server name. In that case the connexion URL should be adapted to the serverName accessed. As example when the user access to https://servername.edu/uPortal the cas auth url is like https://cas.servername.edu/cas/login?service=https://servername.edu/uPortal/login.
But on an other servername like https://othername.edu/uPortal the service should change to https://cas.servername.edu/cas/login?service=https://othername.edu/uPortal/login.

So this PR:
- Provide a configurable way to set the current uPortal serverName to cas service URL:
- Providing also a decorator to be able to set WAYF parameters depending on condition like current serverName. I'm using a such decorator on some serverName where we need to bypass the WAYF. or to add a generated token (this token generator is not provided as internal, but could be added by bean configuration)

[individual contributor license agreement]: https://github.com/Jasig/uPortal/blob/master/CONTRIBUTING.md#individual-contributor-license-agreement
[commit guidelines]: https://github.com/Jasig/uPortal/blob/master/CONTRIBUTING.md#commit
[message properties]: https://github.com/Jasig/uPortal/tree/master/uportal-war/src/main/resources/properties/i18n
[WCAG 2.0 AA]: https://www.w3.org/WAI/WCAG20/quickref/?levels=aaa&technologies=smil%2Cpdf%2Cflash%2Csl

### CODE REVIEW

File: uPortal-security/uPortal-security-mvc/src/test/java/org/apereo/portal/url/CasLoginRefUrlEncoderTest.java [hunk 1]
```diff
@@ -16,12 +16,15 @@
...
+import com.google.common.collect.Sets;
 import java.io.UnsupportedEncodingException;
+import java.util.ArrayList;
+import java.util.List;
 import org.junit.Before;
 import org.junit.Test;
 import org.springframework.mock.web.MockHttpServletRequest;
 
-/** @Author James Wennmacher, jwennmacher@unicon.net */
+/** */
```
Comment (ChristianMurphy) [inline uPortal-security/uPortal-security-mvc/src/test/java/org/apereo/portal/url/CasLoginRefUrlEncoderTest.java]: :art: empty doc comment?

File: uPortal-url/src/main/java/org/apereo/portal/url/IAuthUrlCustomizer.java [hunk 2]
```diff
@@ -0,0 +1,22 @@
+package org.apereo.portal.url;
+
+import javax.servlet.http.HttpServletRequest;
+
+/** Customize CAS or any other authentication URL. */
+public interface IAuthUrlCustomizer {
+    /**
+     * Does this customizer supports the current HTTP request.
+     *
+     * @param request
```
Comment (ChristianMurphy) [inline uPortal-url/src/main/java/org/apereo/portal/url/IAuthUrlCustomizer.java]: `request` but no `url` param?

File: uPortal-url/src/main/java/org/apereo/portal/url/IAuthUrlCustomizer.java [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+     * @return
+     */
+    boolean supports(HttpServletRequest request, String url);
+
+    /**
+     * Customize the supplied external login URL.
+     *
+     * @param url
```
Comment (ChristianMurphy) [inline uPortal-url/src/main/java/org/apereo/portal/url/IAuthUrlCustomizer.java]: `url` but no `request` param?

File: uPortal-security/uPortal-security-mvc/src/test/java/org/apereo/portal/url/CasLoginRefUrlEncoderTest.java (hunk 1 above)
Comment (jgribonvald) [reply uPortal-security/uPortal-security-mvc/src/test/java/org/apereo/portal/url/CasLoginRefUrlEncoderTest.java]: @Author is not requested it seems, so what do you want that I add

Comment (ChristianMurphy) [reply uPortal-security/uPortal-security-mvc/src/test/java/org/apereo/portal/url/CasLoginRefUrlEncoderTest.java]: Or remove the comment entirely

Comment (jgribonvald) [reply uPortal-security/uPortal-security-mvc/src/test/java/org/apereo/portal/url/CasLoginRefUrlEncoderTest.java]: I pushed a change and added a comment, but several other test class have a such empty comment, that's why I did it...

File: uPortal-url/src/main/java/org/apereo/portal/url/IAuthUrlCustomizer.java (hunk 3 above)
Comment (jgribonvald) [reply uPortal-url/src/main/java/org/apereo/portal/url/IAuthUrlCustomizer.java]: The 2 methods have the sames params, but each one had a different param :laughing: 
=> Solved

for notes I only moved the working code from my old previous PR, applied it on my portal instance (without a lot of change) and finally come back to push it, so sorry i didn't take the time to look more on the code.

**Comment by jgribonvald** [comment]
I have to watch on how to get processed the login url into oidc issuer. Any idea ? I would need the curent request or maybe registering into auth session object the URL.

**Comment by ChristianMurphy** [comment]
I'm not sure I quite understand the use case you are going for.
My best guess that you want the url access the front end code, for that including the Url as a user attribute and custom claim would work.

**Comment by jgribonvald** [comment]
The problem with my commit is that issuer have the default url from uPortal.properties, the URL isn't processed by my Customizer and so the url isn't good into oidc attributes like the issuer (which is set from uPortal.properties)

**Comment by jgribonvald** [comment]
I've updated the commit informations about the use case.

**Pull request merged**